### PR TITLE
[SPARK-15762][SQL] Cache Metadata & StructType hashCodes; use singleton Metadata.empty

### PR DESCRIPTION
**Pull request by JoshRosen**
We should cache `Metadata.hashCode` and use a singleton for `Metadata.empty` because calculating metadata hashCodes appears to be a bottleneck for certain workloads.

We should also cache `StructType.hashCode`.

In an optimizer stress-test benchmark run by @ericl, these `hashCode` calls accounted for roughly 40% of the total CPU time and this bottleneck was completely eliminated by the caching added by this patch.

### CODE REVIEW

**Comment by rxin** [comment]
Can you put some data about performance improvements on the pr description?

**Comment by JoshRosen** [comment]
@rxin, I don't have the specific performance numbers handy but in an optimizer stress-test benchmark run by @ericl, these hashCode calls accounted for roughly 40% of the total CPU time and this bottleneck was completely eliminated by the caching added by this patch.

**Comment by ericl** [comment]
[hash code profile](https://github.com/apache/spark/files/298642/hashcode.svg.zip)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Metadata.scala [hunk 1]
```diff
@@ -104,7 +104,8 @@ sealed class Metadata private[types] (private[types] val map: Map[String, Any])
     }
   }
 
-  override def hashCode: Int = Metadata.hash(this)
+  private lazy val _hashCode: Int = Metadata.hash(this)
+  override def hashCode: Int = _hashCode
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Metadata.scala]: any reason why this is not just
```
override lazy val hashCode: Int = Metadata.hash(this)
```
?

**Comment by rxin** [comment]
LGTM other than that small comment.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Metadata.scala (hunk 1 above)
Comment (JoshRosen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Metadata.scala]: Making `hashCode` into a `lazy val` triggers compilation errors and triggers Scalastyle's "You should implement equals and hashCode together" error; it also causes problems in Scala code which calls `.hashCode()`:

```
[error] /Users/joshrosen/Documents/spark/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala:247: type mismatch;
[error]  found   : AttributeReference.this.metadata.type (with underlying type org.apache.spark.sql.types.Metadata)
[error]  required: ?{def hashCode(): ?}
[error] Note that Metadata extends Any, not AnyRef.
[error] Such types can participate in value classes, but instances
[error] cannot appear in singleton types or in reference comparisons.
[error]     h = h * 37 + metadata.hashCode()
[error]                  ^
[error] /Users/joshrosen/Documents/spark/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala:247: Int does not take parameters
[error]     h = h * 37 + metadata.hashCode()
[error]                                   ^
[error] two errors found
[error] /Users/joshrosen/Documents/spark/sql/catalyst/src/main/scala/org/apache/spark/sql/types/Metadata.scala:41:13: You should implement equals and hashCode together
```

Comment (JoshRosen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Metadata.scala]: If you grep through the codebase we don't use `lazy val hashCode` anywhere else, so I'm going to leave this as-is.

**Comment by JoshRosen** [comment]
Merging to master and 2.0.

**Pull request closed without merging**